compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -299,29 +299,35 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             }
             (&ty::Dynamic(ref data_a, ..), &ty::Dynamic(ref data_b, ..)) => {
                 let val = self.read_immediate(src)?;
+                let (old_data, old_vptr) = val.to_scalar_pair()?;
+                let old_vptr = self.scalar_to_ptr(old_vptr)?;
                 if data_a.principal_def_id() == data_b.principal_def_id() {
                     return self.write_immediate(*val, dest);
                 }
                 // trait upcasting coercion
-                let vptr_entry_idx = self.tcx.vtable_trait_upcasting_coercion_new_vptr_slot((
+                let Some(vptr_entry_idx) = self.tcx.vtable_trait_upcasting_coercion_new_vptr_slot((
                     src_pointee_ty,
                     dest_pointee_ty,
-                ));
-
-                if let Some(entry_idx) = vptr_entry_idx {
-                    let entry_idx = u64::try_from(entry_idx).unwrap();
-                    let (old_data, old_vptr) = val.to_scalar_pair()?;
-                    let old_vptr = self.scalar_to_ptr(old_vptr)?;
-                    let new_vptr = self
-                        .read_new_vtable_after_trait_upcasting_from_vtable(old_vptr, entry_idx)?;
-                    self.write_immediate(Immediate::new_dyn_trait(old_data, new_vptr, self), dest)
-                } else {
-                    self.write_immediate(*val, dest)
-                }
+                )) else {
+                    return self.write_immediate(*val, dest);
+                };
+
+                let (ty, _) = self.get_ptr_vtable(old_vptr)?;
+                let Some(ty::VtblEntry::TraitVPtr(new_trait)) = self.get_vtable_entries(old_vptr)?.get(vptr_entry_idx) else {
```

Comment:
I have not found away to go directly from "here are the old and new traits I am casting between, and the old type" to "here is the vtable you should use", so instead I go through `vtable_trait_upcasting_coercion_new_vptr_slot`. This seems a bit silly. But at least `get_vtable_entries` is a typed and high-level way to access the vtable, I don't need to go through its in-memory representation.

Reply:
If you have a trait object, casting to another trait object (eg `&a as dyn Foo as dyn Baz` where `trait Foo: Bar + Baz {}`) potentially requires accessing the vtable to load a pointer to the new vtable from the old vtable.

Reply:
I know why we have to access the vtable at runtime.

But here, shouldn't it work to just say "give me the vtable for type `ty`, trait `new_trait`"? It doesn't matter which trait we are coming from, right?

Reply:
Right, I think it would be possible to use `self.get_vtable_ptr(ty, data_b)` here, copying what the concrete type -> trait object code does, except getting `ty` from the old vtable rather than `src_pointee_ty`.

Reply:
That seems to work, nice. :)